### DISCUSSION TITLE
Add support for pre-releases

### DISCUSSION THREAD

**Discussion by CaseyHofland** [bewuethr/release-tracker-action#52] [Ideas]
I'm thinking of extending support for pre-releases. e.g. when a user creates a release called `1.2.4-pre.8` it will be picked up by the action.

Presumably, the regexp in 'gettags()' would need to be changed a little and everything would work.

```
# List all semver tags and sort them; exclude pre-release tags
gettags() {
	local prefix=$1
	git tag --list --sort='version:refname' "$prefix*.*.*" \
		| grep --extended-regexp "^$prefix([[:digit:]]+\.){2}[[:digit:]]+$"
}
```

A bit of research would need to be done to learn about github pre-release versioning and what is and isn't a valid pre-release name.

---

It might also be interesting to provide a regex user input as a catch-all, for people who know what they're doing and want to customize their versioning to their needs. But that's way overkill for this discussion.

**Comment by CaseyHofland** [comment]
Does the above sound like an idea you would be interested in? If so I can PR my changes and keep using your action, I would rather not maintain my personal repo on this.

**Comment by CaseyHofland** [comment]
This would be the packing order:
`1.2.4`
`1.2.3`
`1.2.0`
`1.2.4-pre.8`
`1.2.4-pre.5`
`1.2.4-pre`
`1.2.3-pre.6`
`1.2.0-pre`
`1.2.4-exp.2`
`1.2.4-exp`

`1.2` would target the latest of these versions it could find. That should hopefully clarify that I mean to support “latest stable version”

Similarly, `latest` would still point to `1.1.0` if no stable `1.2.x` version could be found.

There also needs to be an input to allow users to set precedence of their pre-releases. Something like: [ ‘pre’, ‘exp’ ] should specify that those are expected pre-release tags. It would then work with multiple workflows (e.g. [ ‘beta’, ‘alpha’ ] and omitting this step will retain current behavior. So likely a good solution all around.

**Comment by bewuethr** [comment]
I'm open to adding support for pre-releases, thank you for creating the issue!

I have a few questions:

- How do you envision this to look like for the "end end user", i.e., the consumers of the action that moves their tags around using release-tracker-action? Currently, I can use
  - `@latest` and get the latest tag that looks like `vX.Y.Z` (if `update-latest` is enabled)
  - `@v1` and get the latest tag that looks like `v1.Y.Z`
  - `@v1.1` and get the latest tag that looks like `v1.1.Z` (unless `update-minor` is disabled)
  
  Where would the pre-release come into this? If I use `@v1`, might I get something like `v1.1.0-alpha`? I think you address that with your comment about "latest stable", but I'm not sure I understand.

- SemVer is very precise about [precedence](https://semver.org/#spec-item-11). Specifically, for pre-releases:

  > Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:
  >
  > 1. Identifiers consisting of only digits are compared numerically.
  > 2. Identifiers with letters or hyphens are **compared lexically in ASCII sort order**.
  > 3. Numeric identifiers always have lower precedence than non-numeric identifiers.
  > 4. A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.

  This means that your list would be sorted like this:

  ```none
  1.2.4
  1.2.4-pre.8
  1.2.4-pre.5
  1.2.4-pre
  1.2.4-exp.2
  1.2.4-exp
  1.2.3
  1.2.3-pre.6
  1.2.0
  1.2.0-pre
  ```

  (Annoyingly, GNU `sort -V` does *not* sort like that.)

  The ordering of `exp` and `pre` wouldn't be up to the maintainer, but by ASCII sort order. 

What do you think about this:

1. A new global tag that behave like `latest` (if enabled), but it points to the latest tag overall, including pre-releases. It would be called something like `latest-pre`.
2. Major version tags that behave like the existing ones, but include pre-releases, called something like `v1-pre`, `v2-pre`, etc.
3. If `update-minor` isn't disabled, minor version tags that behave like the existing ones, but include  pre-releases, called something like `v1.1-pre`, `v1.2-pre`, etc.

So, for a list of tags like

```none
v2.1.1-alpha
v2.1.0
v2.0.0
v2.0.0-alpha
v1.1.0-alpha
v1.0.0
v0.1.0
```

the generated tags would be

```none
v2.1.1-alpha  | latest-pre v2-pre v2.1-pre
v2.1.0        | latest v2 v2.1
v2.0.0        | v2.0 v2.0-pre
v2.0.0-alpha  |
v1.1.0-alpha  | v1-pre v1.1-pre
v1.0.0        | v1 v1.0 v1.0-pre
v0.1.0        | v0 v0.1 v0.1-pre
```

What do you think?

**Reply by CaseyHofland** [reply]
Thanks for bringing the sorting to my attention. While it may not be sorted the way I imagined, I think the precedence still stands.

So maybe the best solution would be as follows:
```
sort versions
find first stable version
if no stable version use first version
```

In a list like this:
```
resolve for v2, v2.1, v2.2

v2.2.0-beta   | v2.2
v2.2.0-alpha  |
v2.1.1-alpha  | 
v2.1.0        | v2, v2.1
v2.0.0        |
v2.0.0-alpha  |
```

Versions would then resolve like this:
```
0.1.0
0.1.1
0.1.2
0.1.3         <-- 0.1
0.1.4-beta
0.2.0
0.2.1
0.2.2
0.2.3
0.2.4         <-- 0.2 <-- 0
0.3.0-beta
0.3.0-beta.2  
0.3.1-alpha   <-- 0.3
1.0.0
1.0.1         <-- 1.0
1.1.0
1.1.1
1.1.2         <-- 1.1 <-- 1 <-- latest
1.2.1-alpha
1.2.1-alpha.2
1.2.2-alpha   <-- 1.2
```

**Reply by bewuethr** [reply]
In your lists, why does 1.2 point to a pre-release, but 1 and latest don't?

**Reply by CaseyHofland** [reply]
Either way having tags like `latest-pre`, `v2.1-pre` wouldn't help me meaningfully. The reason for wanting to support pre-release tags is so that, when the pre-release goes into official release, users don't have to update their dependency. They simply point towards `1.2` and once it gets out of alpha, no further action on their side is needed.

It would allow them a sort of *fast track* if they are ok with api that is still experimental to get the latest features.

**Reply by CaseyHofland** [reply]
`1.1`, `1` and `latest` can still find a stable version, while `1,2` cannot so it resolves to the highest unstable version.

**Reply by bewuethr** [reply]
But why would the 1.2 tag even exist before a stable 1.2.X version has been tagged?

The idea behind the `-pre` tags would be that as a user of an action, I can decide if I want just the stable versions (v1, v1.1, latest), or the latest version including pre-releases (v1-pre, v1.1-pre, latest-pre). If the latest version overall is stable, the `-pre` tags would point to that, too, so users wouldn't have to switch between types of moving tags.

**Reply by CaseyHofland** [reply]
The `1.2` tag would exist before a stable `1.2.X` version has been tagged especially for the reason that you can get fast track support for `1.2`, and it would automatically transition into stable support once a stable version releases.

The reason this might be preferable is when e.g. developing packages side by side. You don't want to release the latest version yet, but you do want to reference the latest pre-release in another package.

Having a differentiator for this e.g. `1.2-pre`, `1-pre`, `latest-pre` may bloat your available tags, making it harder for users to decide what version to get. But I do see the concern of not communicating to users if they are on a pre-release version.

**Reply by CaseyHofland** [reply]
I may have a middle ground solution, not bloating the tags but also being very outspoken about what version you're getting.

```
v2.1.1-alpha  | v2.1-pre
v2.1.0        | latest v2 v2.1
v2.0.0        | v2.0
v2.0.0-alpha  | v2.0-pre

v1.1.0        | v1 v1.1
v1.1.0-beta   | v1.1-pre
v1.1.0-alpha.3|
v1.1.0-alpha  |
v1.0.0        | v1.0
```

This would only create minor pre-release versions, which is a good thing, because you don't want to reference `v1-pre` and find yourself in this situation:

```
v1.5.3-alpha.1 | v1-pre now references this...
v1.1.0-beta.8  | v1-pre used to referenced this...
```

That would be a nightmare.

However, it does allow you to specify a pre-release version. Presumably someone who's playing with the pre-release version will keep a close eye on its release status. When that happens they can easily update their reference of `v1.1-pre` to either `v1.1` or `v1` depending on their preference.

**Reply by bewuethr** [reply]
The use case I had in mind when creating this was for Actions, are you using moving tags for something else?

**Reply by CaseyHofland** [reply]
I want to use it for Unity packages, which follow the same structure as code libraries. I keep a lot of my systems separate in order to ensure quality, generalized solutions and extendibility.

It’s common for me to want to pre-release some new functionality in my core package so that I can use it in the development for other packages, then release it when it’s ready.

**Reply by bewuethr** [reply]
Ah, I see. My motivation to create this action was for actions, to mimic the behaviour of the GitHub-provided action such as checkout, cache etc., all of which provide a major version tag that moves to the most recent release within that version. Dependabot even understands that and keeps the "level of precision" the same when upgrading actions, i.e., it would never suggesting to upgrade `v1` to `v2.0.0`, but `v2` instead.

Because of this, tag bloat isn't really a concern: end users of the action don't care if there are a ton of tags in the repository hosting the action. This might be different for your use case, though.

Why are minor versions treated different from the rest in your proposal?

**Reply by CaseyHofland** [reply]
See what I wrote here:
https://github.com/bewuethr/release-tracker-action/discussions/52#discussioncomment-8027183

To clarify, major pre-release versions are usually not even remotely similar so there’s really nothing to gain from it. Minor pre-release versions might change their api but it’s still in service of the same feature.

**Comment by CaseyHofland** [comment]
I've been playing a little with how this would look like on the yaml side of things.

Seeing how you can just use sort -V the idea of providing an input to specify pre-release versions seems silly now.

By relying on ASCII sorting, the following would just work:
```yaml
on:
  push:
    tags:
      # Switch to '[0-9]+.[0-9]+.[0-9]+' if prepend-v is false
      - v[0-9]+.[0-9]+.[0-9]+
      # To support pre-release versions, add the following lines
      - v[0-9]+.[0-9]+.[0-9]+-alpha
      - v[0-9]+.[0-9]+.[0-9]+-alpha.[0-9]+
      - v[0-9]+.[0-9]+.[0-9]+-beta
      - v[0-9]+.[0-9]+.[0-9]+-beta.[0-9]+
```

**Reply by bewuethr** [reply]
To be clear, `sort -V` behaviour deviates from the SemVer spec, so it can't be used to sort tags, I'd say.

Also, notice that the patterns under `tags` are [glob patterns](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushbranchestagsbranches-ignoretags-ignore), not regular expressions, so `[0-9]+` wouldn't work.

**Reply by CaseyHofland** [reply]
I don't know what to tell ya, they work fine.

![Working](https://github.com/bewuethr/release-tracker-action/assets/27729987/54297731-6e04-45d1-b0a7-72da76f7ebab)

Obviously it doesn't do anything, but it certainly starts the action.

**Reply by bewuethr** [reply]
Oh, you're right, the patterns are a strange mix of file globs and regex (https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet)

**Comment by CaseyHofland** [comment]
I made a PR with a possible solution. #53 

It will create minor pre-release tags that will not update once the feature goes out of pre-release. Weather this is preferable remains a part of this discussion.

Example:
```text
v2.1.1-alpha  | v2.1-pre
v2.1.0        | latest v2 v2.1
v2.0.0        | v2.0
v2.0.0-alpha  | v2.0-pre

v1.1.0        | v1 v1.1
v1.1.0-beta   | v1.1-pre
v1.1.0-alpha.3|
v1.1.0-alpha  |
v1.0.0        | v1.0
```

**Reply by bewuethr** [reply]
Thanks, I'll take a look in the next days! I've got to fix the linter, which breaks for PRs from forks...

**Comment by bewuethr** [comment]
This functionality is now released: https://github.com/bewuethr/release-tracker-action/releases/tag/v1.6.0
